src/gui_common/TweakedColourPicker.cs

```diff
@@ -0,0 +1,404 @@
+        rawCheckButton.Disabled = !rawButtonEnabled || HsvMode;
+    }
+
+    private void OnAddPresetButtonPressed()
+    {
+        if (!presetsEnabled)
+            return;
+
+        AddPreset(Color);
+    }
+
+    private void OnHSVButtonToggled(bool isOn)
+    {
+        HsvMode = isOn;
+    }
+
+    private void OnRawButtonToggled(bool isOn)
+    {
+        RawMode = isOn;
+    }
+
+    private void OnColourChanged(Color colour)
+    {
+        htmlColourEdit.Text = colour.ToHtml();
+    }
+
+    /// <summary>
+    ///   Called when (keyboard) entered in HtmlColourEdit or from OnHtmlColourEditFocusExited.
+    ///   Set Color when text is valid; reset if not.
+    /// </summary>
+    /// <param name="colour">Current htmlColourEditor text</param>
+    private void OnHtmlColourEditEntered(string colour)
+    {
+        if (colour.IsValidHtmlColor())
+        {
+            Color = new Color(colour);
+        }
+        else
+        {
+            htmlColourEdit.Text = Color.ToHtml();
+        }
+    }
+
+    /// <summary>
+    ///   Called when focus exited HtmlColourEdit.
+    /// </summary>
+    private void OnHtmlColourEditFocusExited()
+    {
+        OnHtmlColourEditEntered(htmlColourEdit.Text);
+        htmlColourEdit.Deselect();
+    }
+
+    private class TweakedColourPickerPreset : ColorRect
+    {
+        private readonly TweakedColourPicker owner;
+
+        public TweakedColourPickerPreset(TweakedColourPicker owner, Color colour)
+        {
+            this.owner = owner;
+            Color = colour;
+            MarginTop = MarginBottom = MarginLeft = MarginRight = 6.0f;
+            RectMinSize = new Vector2(20, 20);
+            SizeFlagsHorizontal = 8;
+            SizeFlagsVertical = 4;
+            UpdateTooltip();
+        }
+
+        public override void _Notification(int what)
+        {
+            if (what == NotificationTranslationChanged)
+                UpdateTooltip();
+
+            base._Notification(what);
+        }
+
+        public override void _GuiInput(InputEvent inputEvent)
+        {
+            if (inputEvent is InputEventMouseButton { Pressed: true } mouseEvent)
+            {
+                switch ((ButtonList)mouseEvent.ButtonIndex)
+                {
+                    case ButtonList.Left:
+                        owner.Color = Color;
+                        break;
+                    case ButtonList.Right:
+                        if (!owner.PresetsEnabled)
+                            break;
+
+                        owner.ErasePreset(Color);
+                        break;
+                }
+            }
+
+            base._GuiInput(inputEvent);
+        }
+
+        private void UpdateTooltip()
+        {
+            HintTooltip = string.Format(CultureInfo.CurrentCulture,
+                TranslationServer.Translate("COLOUR_PICKER_PRESET_TOOLTIP"), Color.ToHtml());
+        }
+    }
+
+    private class PresetGroupStorage
```

Comment:
So for this class I'd make this implement IEnumerable<Color>, making the Colours a private list and adding two methods: Add and Remove for the colour, which would handle the Colours list, and also notifying the AddPreset and ErasePreset callbacks.

Also one more thing: usually callback calls look like `AddPreset?.Invoke(...)` why that form is not used in this class?

Reply:
Wouldn't [ObservableCollection](https://docs.microsoft.com/en-us/dotnet/api/system.collections.objectmodel.observablecollection-1?view=net-5.0) be a perfect fit for this?

Reply:
Probably yes, I'd say.

Reply:
I'll try (and learn).